uaclient/security.py

```diff
@@ -465,69 +498,109 @@ def prompt_for_affected_packages(
+    if not any([upgrade_packages, upgrade_packages_ua]):
+        print(status.MESSAGE_SECURITY_UPDATE_INSTALLED)
+        print(status.MESSAGE_SECURITY_ISSUE_RESOLVED.format(issue=issue_id))
+    elif upgrade_packages_and_attach(
+        cfg, upgrade_packages, upgrade_packages_ua
+    ):
+        print(status.MESSAGE_SECURITY_UPDATE_INSTALLED)
+        print(status.MESSAGE_SECURITY_ISSUE_RESOLVED.format(issue=issue_id))
+
+
+def _prompt_for_attach(cfg: UAConfig) -> bool:
+    """Prompt for attach to a subscription or token.
+
+    :return: True if attach performed.
+    """
+    import argparse
+    from uaclient import cli
+
+    cloud_type = get_cloud_type()
+    if cloud_type in PRO_CLOUDS:
+        print(
+            status.MESSAGE_SECURITY_USE_PRO_TMPL.format(
+                title=CLOUD_TYPE_TO_TITLE.get(cloud_type), cloud=cloud_type
+            )
+        )
+    print(status.MESSAGE_SECURITY_UPDATE_NOT_INSTALLED_SUBSCRIPTION)
+    choice = util.prompt_choices(
+        "Choose: [S]ubscribe at ubuntu.com [A]ttach"
+        " existing token [C]ancel",
+        valid_choices=["s", "a", "c"],
+    )
+    if choice == "c":
+        return False
+    if choice == "a":
+        print(status.PROMPT_ENTER_TOKEN)
+        token = input("> ")
+        print(status.colorize_commands([["ua", "attach", token]]))
+        return bool(
+            0
+            == cli.action_attach(
+                argparse.Namespace(token=token, auto_enable=True), cfg
+            )
+        )
+    elif choice == "s":
+        print(status.PROMPT_UA_SUBSCRIPTION_URL)
+        print("TODO: GH: #1413: magic subscription attach")
+    return True
 
 
 def upgrade_packages_and_attach(
     cfg: UAConfig,
     upgrade_packages: "List[str]",
     upgrade_packages_ua: "List[str]",
-):
+) -> bool:
     """Upgrade available packages to fix a CVE.
 
     Upgrade all packages in Ubuntu standard updates regardless of attach
     status.
 
     For UA upgrades, prompt regarding system attach prior to upgrading UA
     packages.
-    """
-    if upgrade_packages:
-        if os.getuid() != 0:
-            print(status.MESSAGE_SECURITY_APT_NON_ROOT)
-            return
-
-        apt.run_apt_command(
-            cmd=["apt-get", "update"],
-            error_msg=status.MESSAGE_APT_UPDATE_FAILED,
-            print_cmd=True,
-        )
 
-        apt.run_apt_command(
-            cmd=["apt-get", "install", "--only-upgrade", "-y"]
-            + upgrade_packages,
-            error_msg=status.MESSAGE_APT_INSTALL_FAILED,
-            env={"DEBIAN_FRONTEND": "noninteractive"},
-            print_cmd=True,
-        )
+    :return: True if package upgrade completed or unneeded, False otherwise.
+    """
+    if not any([upgrade_packages, upgrade_packages_ua]):
+        return True
+    if os.getuid() != 0:
+        print(status.MESSAGE_SECURITY_APT_NON_ROOT)
+        return False
     if upgrade_packages_ua:
```

Comment:
Per @mitechie's thoughts, since we can't get to a full fix, we don't want the complexity of a half-fixed solution here at the moment.  I filed https://github.com/canonical/ubuntu-advantage-client/issues/1423 so we can track this separately.